File: backend/routes/records.js
import { Router } from "express";
import db from "../db.js";
import { resolveLocation } from "../services/geocode.js";
import { getTemperaturesForRange, validateDateRange } from "../services/weather.js";

const router = Router();

/**
 * CREATE - POST /api/records
 * Body: { location, startDate, endDate, notes? }
 * Validates the date range, fuzzy-resolves the location, fetches real
 * temperatures for the range, and persists everything.
 */
router.post("/", async (req, res, next) => {
  try {
    const { location, startDate, endDate, notes = "" } = req.body || {};

    const dateError = validateDateRange(startDate, endDate);
    if (dateError) return res.status(400).json({ error: dateError });

    const place = await resolveLocation(location); // throws 404 if not found
    const temps = await getTemperaturesForRange(place.lat, place.lon, startDate, endDate);

    const result = db
      .prepare(
        `INSERT INTO weather_records
         (location_query, resolved_name, latitude, longitude, start_date, end_date, temperature_data, notes)
         VALUES (?, ?, ?, ?, ?, ?, ?, ?)`
      )
      .run(location.trim(), place.name, place.lat, place.lon, startDate, endDate,
           JSON.stringify(temps), String(notes).slice(0, 500));

    const record = db
      .prepare("SELECT * FROM weather_records WHERE id = ?")
      .get(result.lastInsertRowid);

    res.status(201).json(hydrate(record));
  } catch (err) {
    next(err);
  }
});

/** READ (all) - GET /api/records */
router.get("/", (req, res) => {
  const rows = db
    .prepare("SELECT * FROM weather_records ORDER BY created_at DESC")
    .all();
  res.json(rows.map(hydrate));
});

/** READ (one) - GET /api/records/:id */
router.get("/:id", (req, res) => {
  const row = getRecordOr404(req, res);
  if (row) res.json(hydrate(row));
});

/**
 * UPDATE - PUT /api/records/:id
 * Editable fields: location, startDate, endDate, notes.
 * If location or dates change, data is re-validated and temperatures re-fetched
 * so the stored weather never contradicts the stored location/range.
 * Immutable: id, created_at, temperature_data (always derived, never hand-edited).
 */
router.put("/:id", async (req, res, next) => {
  try {
    const existing = getRecordOr404(req, res);
    if (!existing) return;

    const {
      location = existing.location_query,
      startDate = existing.start_date,
      endDate = existing.end_date,
      notes = existing.notes,
    } = req.body || {};

    const dateError = validateDateRange(startDate, endDate);
    if (dateError) return res.status(400).json({ error: dateError });

    const locationChanged = location.trim() !== existing.location_query;
    const datesChanged = startDate !== existing.start_date || endDate !== existing.end_date;

    let place = {
      name: existing.resolved_name,
      lat: existing.latitude,
      lon: existing.longitude,
    };
    let temps = JSON.parse(existing.temperature_data);

    if (locationChanged) place = await resolveLocation(location);
    if (locationChanged || datesChanged) {
      temps = await getTemperaturesForRange(place.lat, place.lon, startDate, endDate);
    }

    db.prepare(
      `UPDATE weather_records SET
         location_query = ?, resolved_name = ?, latitude = ?, longitude = ?,
         start_date = ?, end_date = ?, temperature_data = ?, notes = ?,
         updated_at = datetime('now')
       WHERE id = ?`
    ).run(location.trim(), place.name, place.lat, place.lon, startDate, endDate,
          JSON.stringify(temps), String(notes).slice(0, 500), existing.id);

    const updated = db
      .prepare("SELECT * FROM weather_records WHERE id = ?")
      .get(existing.id);
    res.json(hydrate(updated));
  } catch (err) {
    next(err);
  }
});

/** DELETE - DELETE /api/records/:id */
router.delete("/:id", (req, res) => {
  const row = getRecordOr404(req, res);
  if (!row) return;
  db.prepare("DELETE FROM weather_records WHERE id = ?").run(row.id);
  res.json({ deleted: true, id: row.id });
});

/* ---------- helpers ---------- */

function getRecordOr404(req, res) {
  const id = Number(req.params.id);
  if (!Number.isInteger(id) || id < 1) {
    res.status(400).json({ error: "Record id must be a positive integer." });
    return null;
  }
  const row = db.prepare("SELECT * FROM weather_records WHERE id = ?").get(id);
  if (!row) {
    res.status(404).json({ error: `No record with id ${id}.` });
    return null;
  }
  return row;
}

/** Convert a DB row to API shape (parse the JSON temperature blob). */
function hydrate(row) {
  return {
    id: row.id,
    locationQuery: row.location_query,
    resolvedName: row.resolved_name,
    latitude: row.latitude,
    longitude: row.longitude,
    startDate: row.start_date,
    endDate: row.end_date,
    temperatures: JSON.parse(row.temperature_data),
    notes: row.notes,
    createdAt: row.created_at,
    updatedAt: row.updated_at,
  };
}

export default router;

File: backend/routes/records.py
"""CRUD for weather records: /api/records"""
import json
from fastapi import APIRouter, Body
from db import get_conn
from errors import ApiError
from services.geocode import resolve_location
from services.weather import get_temperatures_for_range, validate_date_range

router = APIRouter()


def hydrate(row) -> dict:
    return {
        "id": row["id"], "locationQuery": row["location_query"], "resolvedName": row["resolved_name"],
        "latitude": row["latitude"], "longitude": row["longitude"],
        "startDate": row["start_date"], "endDate": row["end_date"],
        "temperatures": json.loads(row["temperature_data"]), "notes": row["notes"],
        "createdAt": row["created_at"], "updatedAt": row["updated_at"],
    }


def _get_or_404(conn, record_id: int):
    if record_id < 1:
        raise ApiError(400, "Record id must be a positive integer.")
    row = conn.execute("SELECT * FROM weather_records WHERE id = ?", (record_id,)).fetchone()
    if not row:
        raise ApiError(404, f"No record with id {record_id}.")
    return row


@router.post("/api/records", status_code=201)
def create_record(body: dict = Body(...)):
    """CREATE: validate dates, fuzzy-resolve location, fetch real temps, persist."""
    location, start, end = body.get("location"), body.get("startDate"), body.get("endDate")
    notes = str(body.get("notes") or "")[:500]
    err = validate_date_range(start, end)
    if err:
        raise ApiError(400, err)
    place = resolve_location(location)
    temps = get_temperatures_for_range(place["lat"], place["lon"], start, end)
    with get_conn() as conn:
        cur = conn.execute(
            """INSERT INTO weather_records
               (location_query, resolved_name, latitude, longitude, start_date, end_date, temperature_data, notes)
               VALUES (?, ?, ?, ?, ?, ?, ?, ?)""",
            (location.strip(), place["name"], place["lat"], place["lon"], start, end, json.dumps(temps), notes))
        row = conn.execute("SELECT * FROM weather_records WHERE id = ?", (cur.lastrowid,)).fetchone()
    return hydrate(row)


@router.get("/api/records")
def list_records():
    """READ all."""
    with get_conn() as conn:
        rows = conn.execute("SELECT * FROM weather_records ORDER BY created_at DESC, id DESC").fetchall()
    return [hydrate(r) for r in rows]


@router.get("/api/records/{record_id}")
def get_record(record_id: int):
    """READ one."""
    with get_conn() as conn:
        return hydrate(_get_or_404(conn, record_id))


@router.put("/api/records/{record_id}")
def update_record(record_id: int, body: dict = Body(...)):
    """UPDATE: editable location/dates/notes; temps re-derived when location/range changes."""
    with get_conn() as conn:
        existing = _get_or_404(conn, record_id)
    location = body.get("location", existing["location_query"])
    start = body.get("startDate", existing["start_date"])
    end = body.get("endDate", existing["end_date"])
    notes = str(body.get("notes", existing["notes"]) or "")[:500]

    err = validate_date_range(start, end)
    if err:
        raise ApiError(400, err)

    location_changed = location.strip() != existing["location_query"]
    dates_changed = start != existing["start_date"] or end != existing["end_date"]
    place = {"name": existing["resolved_name"], "lat": existing["latitude"], "lon": existing["longitude"]}
    temps = json.loads(existing["temperature_data"])
    if location_changed:
        place = resolve_location(location)
    if location_changed or dates_changed:
        temps = get_temperatures_for_range(place["lat"], place["lon"], start, end)

    with get_conn() as conn:
        conn.execute(
            """UPDATE weather_records SET location_query=?, resolved_name=?, latitude=?, longitude=?,
               start_date=?, end_date=?, temperature_data=?, notes=?, updated_at=datetime('now') WHERE id=?""",
            (location.strip(), place["name"], place["lat"], place["lon"], start, end, json.dumps(temps), notes, record_id))
        row = conn.execute("SELECT * FROM weather_records WHERE id = ?", (record_id,)).fetchone()
    return hydrate(row)


@router.delete("/api/records/{record_id}")
def delete_record(record_id: int):
    """DELETE."""
    with get_conn() as conn:
        _get_or_404(conn, record_id)
        conn.execute("DELETE FROM weather_records WHERE id = ?", (record_id,))
    return {"deleted": True, "id": record_id}
